### ISSUE TITLE
Warmup trigger behavior on Dedicated plan scale out

### ISSUE THREAD

**Issue by senthilg0pal** [Azure/azure-functions-host#9737]
#### What language does your question apply to? (e.g. C#, JavaScript, Java, All) 
C#

#### Question
We have http trigger endpoints that serves as public apis and we have warmup trigger as well as to load the dependenciesfor the scale out operations. 
But we are not seeing the correct behavior according to the [doc ](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-warmup?tabs=isolated-process%2Cnodejs-v4&pivots=programming-language-csharp).

According to the documentation, when we do the scale out we expect the warmup trigger to load first and not let any other request come in until the warmup finishes its job. But for us some http requests fires before the warmup in the new instance which makes the initial requests slow because the dependencies are not loaded properly and thus experiencing cold start.

Is there anything we have to do other than creating a warmup trigger or is it not guaranteed that warmup will be the first request for a new instance?

**Comment by empowerNate** [comment]
I don't think warmupTrigger works on dedicated, only on the elastic premium plan. This is not in the docs but is my understanding.

**Comment by senthilg0pal** [comment]
Yes, we could not rely on `warmupTrigger` as we are on dedicated
We found a way to warmup before an instance is ready to serve api requests by utilizing Health Check feature.

**Comment by empowerNate** [comment]
How does that work? Is it cheaper to run on dedicated like that compared with premium?

**Comment by senthilg0pal** [comment]
[Health Check](https://learn.microsoft.com/en-us/azure/app-service/monitor-instances-health-check?tabs=dotnet)

So when using health check feature we can provide an endpoint that does all our pre warming logic. 

I did some experiment and saw that first couple of requests will always be the health check endpoint which helped the instance to be pre warmed before serving actual api requests..

We havent tried out elastic premium though, I remember we considered it and then decided not to use for some reason.

Dedicated is not cheap as well. We are actively looking for options to reduce costs and but also make the apis more reliable e.g.. trying out linux plan instead of windows

**Comment by empowerNate** [comment]
Hmm ok. Yeah I can't quite figure out if elastic or dedicated would be more expensive, or if about the same. I think they are roughly the same.

Good to know on the health check though, thanks.

**Comment by empowerNate** [comment]
It looks like elastic premium is about 2x the cost per vCPU, so maybe that's why.

**Comment by empowerNate** [comment]
@senthilg0pal how did you get that to work? Even with health check and a warmupTrigger I'm still getting cold starts with a few requests on scale out. I set the WEBSITE_HEALTHCHECK_MAXPINGFAILURES to 2 (the min) and healthcheck and the healthcheck/warmup code is calling the same things are prod code so that I can be sure everything is warm and working. I suspect that the healthcheck/warmup is being called on one Python worker, but that doesn't mean all the Python workers are warm, and I'm not sure how to get around this.

**Comment by senthilg0pal** [comment]
oh, I am not sure about python workers. Sorry I could not help.

Our warmup logic is basically where we preload all our dependencies (services, cache etc.,). Unfortunately, you will see some initial request takes time but its better than without any warmup.

I found the Health check funcionality  is similar to the Kubernetes - liveness prob. I wish we had something like kubernetes-readiness probe.

May be if you are interested, take a look at this https://github.com/ExpediaGroup/mittens

**Comment by empowerNate** [comment]
Ah ok, you are using C# or something else. Thanks for all the info here!

**Comment by JAdluri** [comment]
Hello @senthilg0pal could you please let me know if the issue still exists.
